### PR TITLE
[PDI-18630] VFS Process files Copy operation fails sporadically fails…

### PR DESCRIPTION
**Pull request by ppatricio**
… with Unable to get VFS File object for filename

### CODE REVIEW

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 17m 25s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl core
```
---
:ok_hand: All tests passed!
 
Tests run: **4747**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7255/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 18m 1s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl core
```
---
:ok_hand: All tests passed!
 
Tests run: **4747**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7255/2/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: core/src/main/java/org/pentaho/di/core/vfs/KettleVFS.java [hunk 1]
```diff
@@ -132,71 +131,73 @@ public static FileObject getFileObject( String vfsFilename, FileSystemOptions fs
...
+  // If not, we are going to assume it's a file, when no scheme found ( flag as null ), and it only changes if
   // a scheme is provided.
   //
   public static FileObject getFileObject( String vfsFilename, VariableSpace space, FileSystemOptions fsOptions )
     throws KettleFileException {
+
+    //  Protect the code below from invalid input.
+    if ( vfsFilename == null ) {
+      throw new KettleFileException( "Unexpected null VFS filename",
+        new NullPointerException( "KettleVFS.getFileObject() vfsFilename is NULL" ) );
```
Comment (graimundo) [inline core/src/main/java/org/pentaho/di/core/vfs/KettleVFS.java]: I believe IllegalArgumentException is more apropriate.
Also, Why the need to encapsulate the exception in a KettleFileException?

File: core/src/main/java/org/pentaho/di/core/vfs/KettleVFS.java [hunk 2]
```diff
@@ -132,71 +131,73 @@ public static FileObject getFileObject( String vfsFilename, FileSystemOptions fs
...
         }
       }
 
-      String filename = normalizePath( vfsFilename, relativeFilename );
+      fsOptions = getFileSystemOptions( scheme, vfsFilename, space, fsOptions );
 
-      return myGetFileObject( filename );
+      String filename = normalizePath( vfsFilename, scheme );
+
+      return fsOptions != null ? fsManager.resolveFile( filename, fsOptions ) : fsManager.resolveFile( filename );
```
Comment (graimundo) [inline core/src/main/java/org/pentaho/di/core/vfs/KettleVFS.java]: We resolve the file with a normalized filename and with a fsOptions built without the normalized filename. Shouldn't fsOptions use the normalized as well?

File: core/src/main/java/org/pentaho/di/core/vfs/KettleVFS.java [hunk 3]
```diff
@@ -132,71 +131,73 @@ public static FileObject getFileObject( String vfsFilename, FileSystemOptions fs
...
       FileSystemManager fsManager = getInstance().getFileSystemManager();
       int timeOut = TIMEOUT_LIMIT;
-      boolean relativeFilename = true;
-      String[] initialSchemes = fsManager.getSchemes();
-
-      relativeFilename = checkForSchemeProvider( initialSchemes, relativeFilename, vfsFilename, space, fsOptionsForScheme );
+      String[] schemes = fsManager.getSchemes();
 
-      boolean hasScheme = vfsFilename != null && hasScheme( vfsFilename, PROVIDER_PATTERN_SCHEME );
+      String scheme = getScheme( schemes, vfsFilename );
```
Comment (graimundo) [inline core/src/main/java/org/pentaho/di/core/vfs/KettleVFS.java]: Aren't these redundant? Why not use the ones from the while cycle?

File: core/src/main/java/org/pentaho/di/core/vfs/KettleVFS.java (hunk 1 above)
Comment (ppatricio) [reply core/src/main/java/org/pentaho/di/core/vfs/KettleVFS.java]: Changed.

File: core/src/main/java/org/pentaho/di/core/vfs/KettleVFS.java (hunk 3 above)
Comment (ppatricio) [reply core/src/main/java/org/pentaho/di/core/vfs/KettleVFS.java]: Just to ensure we don't use always one thread cycle execution.

File: core/src/main/java/org/pentaho/di/core/vfs/KettleVFS.java (hunk 2 above)
Comment (ppatricio) [reply core/src/main/java/org/pentaho/di/core/vfs/KettleVFS.java]: This was made to keep the same behavior from previous versions.

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 18m 19s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl core
```
---
:ok_hand: All tests passed!
 
Tests run: **4747**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7255/3/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 17m 52s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl core
```
---
:ok_hand: All tests passed!
 
Tests run: **4747**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7255/4/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Pull request merged**